datafusion/core/tests/sql/group_by.rs

```diff
@@ -905,3 +905,220 @@ async fn csv_query_group_by_order_by_avg_group_by_substr() -> Result<()> {
+        Field::new("c3", DataType::Int32, false),
+    ]));
+
+    let data = RecordBatch::try_new(
+        schema.clone(),
+        vec![
+            Arc::new(Int32Array::from(vec![1, 1, 2, 2])),
+            Arc::new(Int32Array::from(vec![10, 20, 30, 40])),
+            Arc::new(Int32Array::from(vec![50, 60, 70, 80])),
+        ],
+    )?;
+
+    let ctx = use_postgres_dialect();
+    ctx.register_batch("test", data)?;
+
+    let sql = "SELECT c1, SUM(c2) FILTER (WHERE c2 >= 20) AS sum_c2, AVG(c3) FILTER (WHERE c3 <= 70) AS avg_c3 FROM test GROUP BY c1";
+    let actual = execute_to_batches(&ctx, sql).await;
+
+    let expected = vec![
+        "+----+--------+--------+",
+        "| c1 | sum_c2 | avg_c3 |",
+        "+----+--------+--------+",
+        "| 1  | 20     | 55.0   |",
+        "| 2  | 70     | 70.0   |",
+        "+----+--------+--------+",
+    ];
+
+    assert_batches_sorted_eq!(expected, &actual);
+    Ok(())
+}
+
+#[tokio::test]
+async fn query_group_by_distinct_with_filter() -> Result<()> {
+    let schema = Arc::new(Schema::new(vec![
+        Field::new("c1", DataType::Int32, false),
+        Field::new("c2", DataType::Int32, false),
+    ]));
+
+    let data = RecordBatch::try_new(
+        schema.clone(),
+        vec![
+            Arc::new(Int32Array::from(vec![1, 1, 2, 2, 2])),
+            Arc::new(Int32Array::from(vec![10, 20, 20, 30, 40])),
+        ],
+    )?;
+
+    let ctx = use_postgres_dialect();
+    ctx.register_batch("test", data)?;
+    let sql = "SELECT c1, COUNT(DISTINCT c2) FILTER (WHERE c2 >= 20) AS distinct_c2_count FROM test GROUP BY c1";
+    let actual = execute_to_batches(&ctx, sql).await;
+    let expected = vec![
+        "+----+-------------------+",
+        "| c1 | distinct_c2_count |",
+        "+----+-------------------+",
+        "| 1  | 1                 |",
+        "| 2  | 3                 |",
+        "+----+-------------------+",
+    ];
+    assert_batches_sorted_eq!(expected, &actual);
+    Ok(())
+}
+
+#[tokio::test]
+async fn query_without_group_by_with_filter() -> Result<()> {
+    let schema = Arc::new(Schema::new(vec![
+        Field::new("c1", DataType::Int32, false),
+        Field::new("c2", DataType::Int32, false),
+    ]));
+
+    let data = RecordBatch::try_new(
+        schema.clone(),
+        vec![
+            Arc::new(Int32Array::from(vec![1, 1, 2, 2, 2])),
+            Arc::new(Int32Array::from(vec![10, 20, 20, 30, 40])),
+        ],
+    )?;
+
+    let ctx = use_postgres_dialect();
+    ctx.register_batch("test", data)?;
+    let sql = "SELECT SUM(c2) FILTER (WHERE c2 >= 20) AS sum_c2 FROM test";
+    let actual = execute_to_batches(&ctx, sql).await;
+    let expected = vec![
+        "+--------+",
+        "| sum_c2 |",
+        "+--------+",
+        "| 110    |",
+        "+--------+",
+    ];
+    assert_batches_sorted_eq!(expected, &actual);
+    Ok(())
+}
+
+// count is special cased by `aggregate_statistics`
```

Comment:
Another test case that I think is important is when the filter is on a different column than the aggregate. 

```
postgres=# create table test as values (1, 10), (2, 20), (3, 30);
SELECT 3
postgres=# select * from test;
 column1 | column2
---------+---------
       1 |      10
       2 |      20
       3 |      30
(3 rows)

postgres=# select sum(column1) FILTER (WHERE column2 < 30) from test;
 sum
-----
   3
(1 row)
```

Reply:
Also, it would be good to test when the filter filters out all rows (aka the input to the aggregate is empty)

```sql
SELECT SUM(c2) FILTER (WHERE c2 >= 20000000) AS sum_c2 FROM test
```